client/src/app/users/user.service.spec.ts

```diff
@@ -59,119 +59,188 @@ describe('User service: ', () => {
-    expect(req.request.params.get('age')).toEqual('37');
+  describe('getUsers()', () => {
+
+    it('calls `api/users` when `getUsers()` is called with no parameters', () => {
+      // Assert that the users we get from this call to getUsers()
+      // should be our set of test users. Because we're subscribing
+      // to the result of getUsers(), this won't actually get
+      // checked until the mocked HTTP request 'returns' a response.
+      // This happens when we call req.flush(testUsers) a few lines
+      // down.
+      userService.getUsers().subscribe(
+        users => expect(users).toBe(testUsers)
+      );
+
+      // Specify that (exactly) one request will be made to the specified URL.
+      const req = httpTestingController.expectOne(userService.userUrl);
+      // Check that the request made to that URL was a GET request.
+      expect(req.request.method).toEqual('GET');
+      // Check that the request had no query parameters.
+      expect(req.request.params.keys().length).toBe(0);
+      // Specify the content of the response to that request. This
+      // triggers the subscribe above, which leads to that check
+      // actually being performed.
+      req.flush(testUsers);
+    });
 
-    req.flush(testUsers);
+    describe('Calling getUsers() with parameters correctly forms the HTTP request', () => {
+      /*
+       * We really don't care what `getUsers()` returns in the cases
+       * where the filtering is happening on the server. Since all the
+       * filtering is happening on the server, `getUsers()` is really
+       * just a "pass through" that returns whatever it receives, without
+       * any "post processing" or manipulation. So the tests in this
+       * `describe` block all confirm that the HTTP request is properly formed
+       * and sent out in the world, but don't _really_ care about
+       * what `getUsers()` returns as long as it's what the HTTP
+       * request returns.
+       *
+       * So in each of these tests, we'll keep it simple and have
+       * the (mocked) HTTP request return the entire list `testUsers`
+       * even though in "real life" we would expect the server to
+       * return return a filtered subset of the users.
+       */
+
+      it('correctly calls api/users with filter parameter \'admin\'', () => {
+        userService.getUsers({ role: 'admin' }).subscribe(
+          users => expect(users).toBe(testUsers)
+        );
+
+        // Specify that (exactly) one request will be made to the specified URL with the role parameter.
+        const req = httpTestingController.expectOne(
+          (request) => request.url.startsWith(userService.userUrl) && request.params.has('role')
+        );
+
+        // Check that the request made to that URL was a GET request.
+        expect(req.request.method).toEqual('GET');
+
+        // Check that the role parameter was 'admin'
+        expect(req.request.params.get('role')).toEqual('admin');
+
+        // req.flush(testUsers);
```

Comment:
Should this be commented out?

Reply:
Absolutely not. Thanks! I'm just fixed that in a new commit.